bond_app/routes.py

```diff
@@ -0,0 +1,248 @@
+config.read("config.ini")
+
+def is_provider(section_name):
+    return section_name != 'sam' and section_name != 'bond_accepted'
+
+def create_provider(provider_name):
+    client_id = config.get(provider_name, 'CLIENT_ID')
+    client_secret = config.get(provider_name, 'CLIENT_SECRET')
+    open_id_config_url = config.get(provider_name, 'OPEN_ID_CONFIG_URL')
+    fence_base_url = config.get(provider_name, 'FENCE_BASE_URL')
+    user_name_path_expr = config.get(provider_name, 'USER_NAME_PATH_EXPR')
+
+    sam_base_url = config.get('sam', 'BASE_URL')
+
+    extra_authz_url_params = {}
+    extra_params_key = 'EXTRA_AUTHZ_URL_PARAMS'
+    if config.has_option(provider_name, extra_params_key):
+        extra_params_raw = config.get(provider_name, extra_params_key)
+        extra_authz_url_params = ast.literal_eval(extra_params_raw)
+
+    open_id_config = OpenIdConfig(provider_name, open_id_config_url, cache_api)
+    oauth_adapter = OauthAdapter(client_id, client_secret, open_id_config, provider_name)
+    fence_api = FenceApi(fence_base_url)
+    sam_api = SamApi(sam_base_url)
+
+    fence_tvm = FenceTokenVendingMachine(fence_api, sam_api, cache_api, refresh_token_store, oauth_adapter,
+                                         provider_name, fence_token_storage.FenceTokenStorage())
+    return BondProvider(fence_tvm, Bond(oauth_adapter,
+                                        fence_api,
+                                        sam_api,
+                                        refresh_token_store,
+                                        fence_tvm,
+                                        provider_name,
+                                        user_name_path_expr,
+                                        extra_authz_url_params))
+
+
+def _get_provider(provider_name):
+    if provider_name in bond_providers:
+        return bond_providers[provider_name]
+    else:
+        raise exceptions.NotFound("provider {} not found".format(provider_name))
+
+
+class BondProvider:
+    def __init__(self, fence_tvm, bond):
+        self.fence_tvm = fence_tvm
+        self.bond = bond
+
+
+routes = Blueprint('bond', __name__, '/')
+
+cache_api = DatastoreCacheApi()
+refresh_token_store = TokenStore()
+
+bond_providers = {section_name: create_provider(section_name)
+                  for section_name in config.sections() if is_provider(section_name)}
+
+authentication_config = authentication.AuthenticationConfig(config.get('bond_accepted', 'AUDIENCE_PREFIXES').split(),
+                                                            config.get('bond_accepted', 'EMAIL_SUFFIXES').split(),
+                                                            os.environ.get('BOND_MAX_TOKEN_LIFE', 600))
+auth = authentication.Authentication(authentication_config, cache_api)
+
+api_routes_base = '/api/link/v1'
+
+
+@routes.route('/')
+def root():
+    return "Bond - Account Linking Service"
+
+
+@routes.route(api_routes_base + '/providers', methods=["GET"])
```

Comment:
We may need to make a global change for all of our URLs to accept trailing slashes:

```suggestion
@routes.route(api_routes_base + '/providers', methods=["GET"], strict_slashes=False)
```

This would fix this URL, but I think there's a way to enable it globally for all all routes.  Legacy Bond seems to accept trailing slashes whereas new Bond will freak.

Reply:
I tried to enable it globally, but wasn't able to get it to work properly. When it was enabled globally, /providers and /providers/ were still treated as two separate endpoints. Since I couldn't quite get it to work, I just enabled it on each endpoint instead.